### PR TITLE
Change hsearch to consider unquoted parameter strings as prefix match

### PR DESCRIPTION
**Pull request by jmarchionatto**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/search/TermHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+	@Test
+	void empty_returns_empty() {
+		assertEquals( Collections.emptySet(), TermHelper.makePrefixSearchTerm(Collections.emptySet()) );
+	}
+
+	@Test
+	void noQuotedSpcedOrStarElements_return_star_suffixed() {
+		Set<String> result = TermHelper.makePrefixSearchTerm(Set.of("abc", "def", "ghi"));
+		assertEquals( Set.of("abc *", "def *", "ghi *"), result );
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/search/TermHelperTest.java:20]: Yes. Forgot to update the test when fixed the code. Done.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4034-align-text-query-syntax-with-hapi-string-search.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+type: fix
+issue: 4034
+title: "With Elastic/Lucene indexing enabled search including `:text` qualifier was considering tokens literally but
+   standard match in HAPI-FHIR considers arguments as prefixes. This has been corrected by making Elastic/Lucene
+   `:text` search also consider arguments as prefixes."
```
Comment (michaelabuckley) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4034-align-text-query-syntax-with-hapi-string-search.yaml:6]:
```suggestion
type: add
issue: 4034
title: "Added support for the `:text` modifier on string search parameters.  This corrects an issue when using Elastic/Lucene indexing enabled where prefix match was used instead."
```

File: hapi-fhir-jpaserver-elastic-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java [hunk 3]
```diff
@@ -416,6 +414,118 @@ public void testResourceCodeTokenSearch() {
...
+		@Test
+		void stringMatchesPrefixAndWhole() {
+			// smit - matches "smit" and "smith"
+
+			String id1 = myTestDataBuilder.createObservation(List.of(
+				myTestDataBuilder.withPrimitiveAttribute("valueString", "John Smith") )).getIdPart();
+			String id2 = myTestDataBuilder.createObservation(List.of(
+				myTestDataBuilder.withPrimitiveAttribute("valueString", "Carl Smit") )).getIdPart();
+
+			List<String> resourceIds = myTestDaoSearch.searchForIds("/Observation?value-string:text=smit");
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-elastic-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java:438]: Can you add a test case making sure that `/Observation?value-string:text=car%20smit` matches id2?

**Review by michaelabuckley — COMMENTED** [review]
Looks good.  Requested another test case.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/search/TermHelperTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	void empty_returns_empty() {
+		assertEquals( Collections.emptySet(), TermHelper.makePrefixSearchTerm(Collections.emptySet()) );
+	}
+
+	@Test
+	void noQuotedSpcedOrStarElements_return_star_suffixed() {
+		Set<String> result = TermHelper.makePrefixSearchTerm(Set.of("abc", "def", "ghi"));
+		assertEquals( Set.of("abc*", "def*", "ghi*"), result );
+	}
+
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/search/TermHelperTest.java:22]: I'm not sure about all the results here, but we need to handle multi-term queries.
```suggestion
	@Test
 	void multiSimpleTerm_hasSimpleTermsWildcarded() {
 		Set<String> result = TermHelper.makePrefixSearchTerm(Set.of("abc def"));
 		assertEquals( Set.of("abc* def*"), result );
 	}

  // question - do we want to wildcard simple terms inside other syntax for consistency?
	@Test
 	void simpleQuerySytax_hasSimpleTermsWildcarded() {
 		Set<String> result = TermHelper.makePrefixSearchTerm(Set.of("(def | efg)", "ghi +(\"abc" \"def\")"));
 		assertEquals( Set.of("(def* | efg*)", "ghi* +(\"abc" \"def\")"), result );
 	}

```

File: hapi-fhir-jpaserver-elastic-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java (hunk 3 above)
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-elastic-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java:438]: Done.

**Review by michaelabuckley — APPROVED** [review]
Nice!

**Pull request merged**